Call `removeRetiredPools` from `monitorStakePools`.

jonathanknowles opened the pull request:
# Issue Number

#2019 

# Overview

PR #2057 provided the [`removeRetiredPools`](https://github.com/input-output-hk/cardano-wallet/blob/fece720cb18d9676c4f1f6b0d3a69a37fab93f1c/lib/core/src/Cardano/Pool/DB.hs#L301) function which, when given an epoch, removes pools from the database that retired **on or before** that epoch.

This PR adjusts [`monitorStakePools`](https://github.com/input-output-hk/cardano-wallet/blob/fece720cb18d9676c4f1f6b0d3a69a37fab93f1c/lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs#L481) to actually **call** [`removeRetiredPools`](https://github.com/input-output-hk/cardano-wallet/blob/fece720cb18d9676c4f1f6b0d3a69a37fab93f1c/lib/core/src/Cardano/Pool/DB.hs#L301).

## Frequency of Garbage Collection

Since garbage collection has a **runtime cost** associated with it, we want to avoid doing it more than necessary, even if the cost is relatively small.

Thus, we only perform garbage collection:

- **on demand**, just before we record a new pool registration or retirement certificate in the database.
- **at most once per epoch**.

## Logging Examples

### A single garbage collection invocation (at most once per epoch)

<details><summary>Click to view log sample</summary>

```hs
[cardano-wallet.pools-engine:Debug:31] [2020-08-31 05:42:24.77 UTC] Performing garbage collection of retired stake pools. Currently in epoch 214. Removing all pools that retired in or before epoch 212.
[cardano-wallet.pools-db:Debug:31] [2020-08-31 05:42:24.77 UTC] Removing pools that retired in or before epoch 212: start
[cardano-wallet.pools-db:Debug:31] [2020-08-31 05:42:24.78 UTC] Removing the following retired pools:
Pool 215622e6 with retirement epoch 212
Pool 5687ecfe with retirement epoch 212
Pool 69d158eb with retirement epoch 212
Pool 7bacd597 with retirement epoch 212
Pool 8e681030 with retirement epoch 212
Pool a6d582d8 with retirement epoch 212
Pool bccf5ce0 with retirement epoch 212
Pool c0b86026 with retirement epoch 212
Pool cfe83240 with retirement epoch 212
Pool f5863386 with retirement epoch 212
[cardano-wallet.pools-db:Notice:31] [2020-08-31 05:42:24.78 UTC] Removing the following pool from the database: 215622e669c0f0c09828ec4100a4c689b1f9c148f542b0192b6ed00d.
[cardano-wallet.pools-db:Notice:31] [2020-08-31 05:42:24.80 UTC] Removing the following pool from the database: 5687ecfeaf19ed1a7ec20a9332ff80c8ad6fc5ace97cfb7276d256a2.
[cardano-wallet.pools-db:Notice:31] [2020-08-31 05:42:24.82 UTC] Removing the following pool from the database: 69d158eb61cbbeff3ae564768eef7c0b52ba8bd33b2301eacf62a24d.
[cardano-wallet.pools-db:Notice:31] [2020-08-31 05:42:24.84 UTC] Removing the following pool from the database: 7bacd59728317961e6ffb20f40f74706954a5341d1159f7e2be32a68.
[cardano-wallet.pools-db:Notice:31] [2020-08-31 05:42:24.85 UTC] Removing the following pool from the database: 8e68103060b32595d4eed0a7749795b155d790b95e0615f9f5618b9f.
[cardano-wallet.pools-db:Notice:31] [2020-08-31 05:42:24.87 UTC] Removing the following pool from the database: a6d582d86cafd861984ec33f3c4bb7ab1e183d0719abc92351dd2dd7.
[cardano-wallet.pools-db:Notice:31] [2020-08-31 05:42:24.89 UTC] Removing the following pool from the database: bccf5ce0d9b3436cd1616686c98f4584b43760704c9585f1f208ffae.
[cardano-wallet.pools-db:Notice:31] [2020-08-31 05:42:24.90 UTC] Removing the following pool from the database: c0b86026e65261a0698e789fc2e0b1cf36c85693dbf1cd516067400d.
[cardano-wallet.pools-db:Notice:31] [2020-08-31 05:42:24.92 UTC] Removing the following pool from the database: cfe83240dd1eb92bac3d739b2550be334f4bb7e4fc11d7f5faeb8769.
[cardano-wallet.pools-db:Notice:31] [2020-08-31 05:42:24.94 UTC] Removing the following pool from the database: f58633867bea26335d12aa95a8b3b1dee5b6994ffd4bac5348375220.
[cardano-wallet.pools-db:Debug:31] [2020-08-31 05:42:24.95 UTC] Removing pools that retired in or before epoch 212: finish
```
</details>

### Lifecycle of a single pool

<details><summary>Click to view log sample</summary>

```hs
[cardano-wallet.pools-engine:Info:37] [2020-08-27 15:35:54.31 UTC] Discovered stake pool registration: Registration of 00008d6a owned by [ed25519_pk1lrrgks5hnpxqqpyhjxedk7u5nxp0ehp9tm9ksx4a7zepzt6ydnw]
...
[cardano-wallet.pools-engine:Info:37] [2020-08-27 15:35:54.85 UTC] Discovered stake pool retirement: Pool 00008d6a with retirement epoch 209
...
[cardano-wallet.pools-engine:Debug:37] [2020-08-27 15:36:16.07 UTC] Performing garbage collection of retired stake pools. Currently in epoch 211. Removing all pools that retired in or before epoch 209.
[cardano-wallet.pools-db:Notice:37] [2020-08-27 15:36:16.08 UTC] Removing the following pool from the database: 00008d6abd6bf0dafbc7f64b68267f5716e1992d0e845eb22d05065b.
```

</details>

jonathanknowles committed:
Extract out function `putPoolCertificates`.

jonathanknowles committed:
Add `MsgDb` constructor to `StakePoolLog`.

jonathanknowles committed:
Add function `garbageCollectPools` and call it within `forward`.

jonathanknowles committed:
Limit the frequency of pool garbage collection.

By only performing garbage collection when we actually have new pool
certificates to add, we avoid paying the cost of garbage collection in
every single slot.

jonathanknowles committed:
Add `MsgStakePoolGarbageCollection` constructor to `StakePoolLog`.

jonathanknowles committed:
Write `MsgStakePoolGarbageCollection` to the log.

When garbage collecting pools.

jonathanknowles committed:
Revise comment for `putPoolCertificates`.

jonathanknowles committed:
Appease the hlint deity.

jonathanknowles commented on lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs:
```diff
@@ -621,6 +668,15 @@ instance ToText StakePoolLog where
+            , toText (poolGarbageCollectionRemovalEpoch info)
+            , ". "
+            , toText bkt
+            ]
```
See if this can be made simpler through use of [`Fmt`](http://hackage.haskell.org/package/fmt-0.6.1.2/docs/Fmt.html).

jonathanknowles committed:
Use record wildcards for `PoolGarbageCollectionInfo`.

jonathanknowles committed:
Ignore bracket value in `toText` for `MsgStakePoolGarbageCollection`.

jonathanknowles committed:
Reduce severity of `MsgStakePoolGarbageCollection` to `Debug`.

We don't need to see any informational messages when there are no pools
to garbage collect.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/2057/files#r476141310

jonathanknowles commented on lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs:
```diff
@@ -577,26 +603,40 @@ monitorMetadata tr gp fetchMetadata DBLayer{..} = forever $ do
         MsgHaltMonitoring{} -> Info
         MsgCrashMonitoring{} -> Error
         MsgRollingBackTo{} -> Info
+        MsgStakePoolGarbageCollection{} -> Info
```
This should probably be lowered to `Debug` in order to avoid spamming the log. In particular, we don't need to see any informational messages when there are no pools to garbage collect.

jonathanknowles replied:
> This should probably be lowered to `Debug` in order to avoid spamming the log. In particular, we don't need to see any informational messages when there are no pools to garbage collect.

Fixed in b1e4d1506c15f1be05fc1ad2416badee69a47529.

jonathanknowles commented on lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs:
```diff
@@ -577,26 +605,40 @@ monitorMetadata tr gp fetchMetadata DBLayer{..} = forever $ do
+    { currentEpoch :: EpochNo
+        -- ^ The current epoch at the point in time the garbage collector
+        -- was invoked.
+    , removalEpoch :: EpochNo
```
**Suggestion:** The name `removalEpoch` seems a little unclear. Perhaps call it `latestRetirementEpoch`?

**Justification:** the name `latestRetirementEpoch` implies a relationship with `retirementEpoch` in `PoolRetirementCertificate`, which is exactly the implication we want.

```suggestion
    , latestRetirementEpoch :: EpochNo
```

jonathanknowles replied:
> **Suggestion:** The name `removalEpoch` seems a little unclear. Perhaps call it `latestRetirementEpoch`?

Fixed in 78e8e04f888c7288166147b63d22fdff6ccdc7c2.

jonathanknowles committed:
Remove one level of `where` nesting.

The nested function definitions did not depend on the outer function, so
it is safe to move them up one level.

jonathanknowles committed:
Rename `removalEpoch` to `latestRetirementEpoch`.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/2057#discussion_r476274161

Anviking commented on lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs:
```diff
@@ -517,31 +525,52 @@ monitorStakePools tr gp nl db@DBLayer{..} = do
+    --
+    garbageCollectPools currentSlot = liftIO $ do
+        currentEpoch <- timeInterpreter nl (epochOf currentSlot)
+        let subtractTwoEpochs = epochPred <=< epochPred
```
What if the current epoch is 0? (Perhaps not technically possible this is called then, but still…)

jonathanknowles replied:
@Anviking wrote:
> What if the current epoch is 0? (Perhaps not technically possible this is called then, but still…)

I'm glad you asked :smile_cat:.

In that case, we will not perform any garbage collection, as `subtractTwoEpochs currentEpoch` will evaluate to `Nothing`. (Since [`epochPred`](https://github.com/input-output-hk/cardano-wallet/blob/fece720cb18d9676c4f1f6b0d3a69a37fab93f1c/lib/core/src/Cardano/Wallet/Primitive/Slotting.hs#L364) has type `EpochNo -> Maybe EpochNo`.)

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs:
```diff
@@ -577,26 +606,41 @@ monitorMetadata tr gp fetchMetadata DBLayer{..} = forever $ do
 
 data StakePoolLog
-    = MsgFollow FollowLog
+    = MsgDb PoolDbLog
```
We already have a logger for the pool database, why is this needed here?

jonathanknowles replied:
@KtorZ 
> We already have a logger for the pool database, why is this needed here?

Good question.

When we call the `removeRetiredPools` function, it expects to be provided with a `Tracer IO PoolDbLog` value. But we call it from a context where we only have a `Tracer IO StakePoolLog`. Therefore, we need to have some way to transform the `Tracer IO StakePoolLog` into a `Tracer IO PoolDbLog`.

Luckily, the `contramap` function allows us to perform this transformation, but it needs to be provided with something that maps a `PoolDbLog` into a `StakePoolLog`, which is where the `MsgDb` constructor comes in.

KtorZ replied:
That seems quite wrong to me. Why isn't `removeRetiredPools` part of the DBLayer already? This creates an inconsistency with the "normal approach" which seems unjustified to me. 

The problem with having logs database embedded in the wrong logger is that it completely defies the purpose of having separate logger components. Now, if I run `--trace-pool-db off`, I still get pool db logs regarding the garbage collection of pools.

jonathanknowles replied:
> Why isn't `removeRetiredPools` part of the DBLayer already?

That's a good question.

The `removeRetiredPools` function depends **only** on the DBLayer **interface**, and not any particular implementation. It only calls `listRetiredPools` and `removePools`. So this function can sit outside of the `DBLayer` interface itself (in the same way that we do for other function that only depend on the interface).

One alternative would be to add a `removeRetiredPools` function to the `DBLayer` type, and then implement this twice (once in the `SQLite` module and another time in the `MVar` module), but reuse the same function behind the scenes.

KtorZ replied:
> One alternative would be to add a removeRetiredPools function to the DBLayer type, and then implement this twice (once in the SQLite module and another time in the MVar module), but reuse the same function behind the scenes.

Yes. That sounds preferable for the consumers of the DBLayer such that things like the DBLog doesn't leak outside of the DBLayer.

jonathanknowles replied:
@KtorZ wrote:
> The problem with having logs database embedded in the wrong logger is that it completely defies the purpose of having separate logger components. Now, if I run --trace-pool-db off, I still get pool db logs regarding the garbage collection of pools.

The pattern I've used for `MsgDb` is also used by `MsgFollow`, which embeds entries from the `FollowLog`:

```hs
data StakePoolLog 
    = MsgDb PoolDbLog 
    | MsgFollow FollowLog
    ...

data FollowLog
    = MsgFollowAction (FollowAction String)   
    | MsgUnhandledException Text  
    | MsgNextBlockFailed ErrGetBlock 
    ...
```

See: https://github.com/input-output-hk/cardano-wallet/blob/9f93d0a98be7c19035dd603149ebf06ac38d7f2f/lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs#L484

Is there a better solution that we can use here?

KtorZ replied:
@jonathanknowles FollowLog is a completely different story, it is not part of any other logger component. There are two components that include a `FollowLog` and these are the `stake-pool-engine` and `wallet-engine` loggers. 

However, the DBLog is already a logger component all by itself. See:

https://github.com/input-output-hk/cardano-wallet/blob/master/lib/shelley/src/Cardano/Wallet/Shelley.hs#L463-L472

jonathanknowles replied:
@KtorZ many thanks for your explanation, that makes sense.

I've made the following changes:
* moved `removeRetiredPools` into the `DBLayer`.
* completely removed the `MsgDb` constructor.

See the following commits:
- 2ed42a2e0b1ba8b9a939a855c6ee40505fd348ff
- 742027cb266cb77e6f7bfd073211bbb84bb404d1

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs:
```diff
@@ -517,31 +525,52 @@ monitorStakePools tr gp nl db@DBLayer{..} = do
+    --
+    -- To avoid any issues with rollback, we err on the side of caution and
+    -- only garbage collect pools that retired /at least/ two epochs before
+    -- the current epoch.
```
What's the rationale for "at least 2". Why not, "at least 3" for example?

jonathanknowles replied:
> What's the rationale for "at least 2". Why not, "at least 3" for example?

If we've **only just** crossed the boundary from epoch **_e_** into epoch **_e_** **+ 1**, this means that blocks from the previous epoch are still within the rollback window. If a rollback does happen, and if the chain we've just been following has been replaced with another chain in which the retirement certificate for some pool **p** is revoked within epoch **_e_** (with a registration certificate that supersedes it), then the retirement for pool **p** would be cancelled.

But assuming the rollback window is guaranteed to be less than one epoch in length, retirements that occurred two epochs ago can no longer be affected by rollback. So if we see a retirement that occurred two epochs ago and was not superseded, it is safe to perform garbage collection of that pool.

KtorZ replied:
Would it be possible to extend the comment with this rationale :) ?

jonathanknowles replied:
> Would it be possible to extend the comment with this rationale :) ?

Certainly!

Fixed in 395a64020d9ed0c92d53d0cd4de2d9a38e29a0f9

Anviking replied:
👍

jonathanknowles committed:
Add rationale to `garbageCollectPools` function.

In this rationale, we explain why we only garbage collect pools that
retired at least two epochs before the current epoch.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/2057#discussion_r477229174

jonathanknowles committed:
Use `RecordWildCards` in `newDBLayer` functions.

Justification:

Some `DBLayer` functions are defined in terms of other `DBLayer`
functions, such as `readPoolLifeCycleStatus`, which is defined in terms
of `readPoolRegistration` and `readPoolRetirement`.

Since mutual recursion between functions defined as fields of a record
is somewhat awkward, we currently define auxilliary functions for each
of the functions that we need to reuse, and then refer to these
auxilliary functions when building the record. However, this creates a
layer of indirection and duplication.

By using `RecordWildCards`, we can instead build a `DBLayer` object in
terms of ordinary functions defined in a `where` clause. Since ordinary
functions can refer to one another using mutual recursion, we no longer
need auxilliary functons, saving a layer of indirection.

jonathanknowles committed:
Remove auxilliary functions from `newDBLayer`.

With the use of `RecordWildCards`, these auxilliary functions are no
longer necessary, as functions in a where clause can refer to one
another directly.

jonathanknowles committed:
Make `removeRetiredPools` a first-class citizen of `DBLayer`.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/2057#discussion_r477253887

jonathanknowles committed:
Remove now-unused `MsgDb` constructor.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/2057#discussion_r477253887

jonathanknowles committed:
Remove nested call to `atomically`.

Nesting a call to `atomically` will result in the nested call blocking
indefinitely.

This fixes:

https://github.com/input-output-hk/cardano-wallet/pull/2057#issuecomment-680825458
https://github.com/input-output-hk/cardano-wallet/pull/2057#issuecomment-680833124

This change also removes bracketing for the `MsgStakePoolGarbageCollection`
constructor of `StakePoolLog`, as there are no longer any child log entries
belonging to the pools engine.

(Other log entries pertaining to garbage collection are emitted by the pool
DB component, as a result of calling the `removeRetiredPools` DB function.)

jonathanknowles committed:
Remove redundant comment.

jonathanknowles committed:
Limit pool garbage collection to once per epoch.

Anviking reviewed (approved):
lgtm
